File: packages/hqtui/src/color.ts
/**
 * Colors are packed into a single 32-bit integer so a cell never needs an object.
 *
 *   0                       -> "terminal default"
 *   0x1000000 | 0xRRGGBB    -> truecolor
 *   0x2000000 | index       -> explicit 256-colour palette index
 */

export type Color = number;

export const DEFAULT_COLOR: Color = 0;
const RGB_FLAG = 0x1000000;
const IDX_FLAG = 0x2000000;

/** Truecolor from 0-255 components. */
export function rgb(r: number, g: number, b: number): Color {
  return RGB_FLAG | ((r & 255) << 16) | ((g & 255) << 8) | (b & 255);
}

/** `hex("#00d7ff")`, `hex("#0df")` or `hex(0x00d7ff)`. */
export function hex(value: string | number): Color {
  if (typeof value === "number") return RGB_FLAG | (value & 0xffffff);
  let s = value.trim().replace(/^#/, "");
  if (s.length === 3) s = s[0] + s[0] + s[1] + s[1] + s[2] + s[2];
  return RGB_FLAG | (Number.parseInt(s, 16) & 0xffffff);
}

/** An explicit xterm-256 palette entry (rarely needed; truecolor is quantized for you). */
export function ansi256(index: number): Color {
  return IDX_FLAG | (index & 255);
}

export function isDefault(c: Color): boolean {
  return c === DEFAULT_COLOR;
}

export function red(c: Color): number {
  return (c >> 16) & 255;
}
export function green(c: Color): number {
  return (c >> 8) & 255;
}
export function blue(c: Color): number {
  return c & 255;
}

/** Mix two colors. `t` of 0 returns `a`, 1 returns `b`. Software alpha — terminals have none. */
export function mix(a: Color, b: Color, t: number): Color {
  if (isDefault(a) || isDefault(b)) return t < 0.5 ? a : b;
  const k = t < 0 ? 0 : t > 1 ? 1 : t;
  return rgb(
    Math.round(red(a) + (red(b) - red(a)) * k),
    Math.round(green(a) + (green(b) - green(a)) * k),
    Math.round(blue(a) + (blue(b) - blue(a)) * k),
  );
}

/** Blend a color over a background at `alpha` (0-1). Used for subtle fills and shadows. */
export function alpha(fg: Color, bg: Color, a: number): Color {
  return mix(bg, fg, a);
}

export function lighten(c: Color, amount = 0.2): Color {
  return mix(c, rgb(255, 255, 255), amount);
}

export function darken(c: Color, amount = 0.2): Color {
  return mix(c, rgb(0, 0, 0), amount);
}

/** Relative luminance, 0-1. */
export function luminance(c: Color): number {
  const f = (v: number) => {
    const x = v / 255;
    return x <= 0.03928 ? x / 12.92 : ((x + 0.055) / 1.055) ** 2.4;
  };
  return 0.2126 * f(red(c)) + 0.7152 * f(green(c)) + 0.0722 * f(blue(c));
}

/** WCAG contrast ratio between two colors (1-21). */
export function contrast(a: Color, b: Color): number {
  const la = luminance(a);
  const lb = luminance(b);
  return (Math.max(la, lb) + 0.05) / (Math.min(la, lb) + 0.05);
}

/**
 * A multi-stop gradient. Returns a sampler: `g(0)` is the first stop, `g(1)` the last.
 *
 *   const heat = gradient(["#00d7ff", "#5fff87", "#ffd75f", "#ff5f5f"]);
 *   heat(0.75);
 */
export function gradient(stops: (Color | string)[]): (t: number) => Color {
  const cols = stops.map((s) => (typeof s === "string" ? hex(s) : s));
  if (cols.length === 0) return () => DEFAULT_COLOR;
  if (cols.length === 1) return () => cols[0];
  return (t: number) => {
    const k = t < 0 ? 0 : t > 1 ? 1 : t;
    const pos = k * (cols.length - 1);
    const i = Math.min(Math.floor(pos), cols.length - 2);
    return mix(cols[i], cols[i + 1], pos - i);
  };
}

/** Sample `n` evenly spaced colors from a gradient. */
export function gradientSteps(stops: (Color | string)[], n: number): Color[] {
  const g = gradient(stops);
  const out: Color[] = [];
  for (let i = 0; i < n; i++) out.push(g(n === 1 ? 0 : i / (n - 1)));
  return out;
}

const CUBE = [0, 95, 135, 175, 215, 255];

function nearestCubeIndex(v: number): number {
  let best = 0;
  let bestD = Infinity;
  for (let i = 0; i < 6; i++) {
    const d = Math.abs(CUBE[i] - v);
    if (d < bestD) {
      bestD = d;
      best = i;
    }
  }
  return best;
}

/** Quantize truecolor to the xterm-256 palette (used when COLORTERM is absent). */
export function to256(c: Color): number {
  if ((c & IDX_FLAG) !== 0) return c & 255;
  const r = red(c);
  const g = green(c);
  const b = blue(c);
  // Grey ramp often beats the cube for desaturated colors.
  if (Math.abs(r - g) < 8 && Math.abs(g - b) < 8) {
    if (r < 8) return 16;
    if (r > 248) return 231;
    return 232 + Math.round(((r - 8) / 247) * 24);
  }
  return 16 + 36 * nearestCubeIndex(r) + 6 * nearestCubeIndex(g) + nearestCubeIndex(b);
}

const BASE16: [number, number, number][] = [
  [0, 0, 0], [205, 49, 49], [13, 188, 121], [229, 229, 16],
  [36, 114, 200], [188, 63, 188], [17, 168, 205], [229, 229, 229],
  [102, 102, 102], [241, 76, 76], [35, 209, 139], [245, 245, 67],
  [59, 142, 234], [214, 112, 214], [41, 184, 219], [255, 255, 255],
];

/** Quantize truecolor to the 16-color palette (last-resort terminals). */
export function to16(c: Color): number {
  if ((c & IDX_FLAG) !== 0) {
    const i = c & 255;
    return i < 16 ? i : to16(from256(i));
  }
  const r = red(c);
  const g = green(c);
  const b = blue(c);
  let best = 7;
  let bestD = Infinity;
  for (let i = 0; i < 16; i++) {
    const [br, bg, bb] = BASE16[i];
    const d = (r - br) ** 2 + (g - bg) ** 2 + (b - bb) ** 2;
    if (d < bestD) {
      bestD = d;
      best = i;
    }
  }
  return best;
}

/** Convert a 256-palette index back to truecolor. */
export function from256(index: number): Color {
  const i = index & 255;
  if (i < 16) {
    const [r, g, b] = BASE16[i];
    return rgb(r, g, b);
  }
  if (i >= 232) {
    const v = 8 + (i - 232) * 10;
    return rgb(v, v, v);
  }
  const n = i - 16;
  return rgb(CUBE[Math.floor(n / 36) % 6], CUBE[Math.floor(n / 6) % 6], CUBE[n % 6]);
}

/** Desaturate towards grey — powers `monochrome` mode. */
export function grayscale(c: Color): Color {
  if (isDefault(c)) return c;
  const v = Math.round(0.299 * red(c) + 0.587 * green(c) + 0.114 * blue(c));
  return rgb(v, v, v);
}

File: ports/python/hqtui/color.py
"""Colors packed into a single 32-bit integer so a cell never needs an object.

::

    0                       -> "terminal default"
    0x1000000 | 0xRRGGBB    -> truecolor
    0x2000000 | index       -> explicit 256-colour palette index

A ``Color`` is a plain ``int`` subclass rather than a dataclass: the framebuffer
holds millions of them per second, and an object per cell is exactly what the
reference implementation exists to avoid.
"""

from __future__ import annotations

import math
from typing import Iterable, Sequence

__all__ = [
    "Color",
    "DEFAULT_COLOR",
    "Gradient",
    "from_256",
    "hex_color",
    "rgb",
    "ansi256",
]

_RGB_FLAG = 0x1000000
_IDX_FLAG = 0x2000000

_CUBE = (0, 95, 135, 175, 215, 255)

_BASE16 = (
    (0, 0, 0), (205, 49, 49), (13, 188, 121), (229, 229, 16),
    (36, 114, 200), (188, 63, 188), (17, 168, 205), (229, 229, 229),
    (102, 102, 102), (241, 76, 76), (35, 209, 139), (245, 245, 67),
    (59, 142, 234), (214, 112, 214), (41, 184, 219), (255, 255, 255),
)


def round_half_up(value: float) -> float:
    """JavaScript's ``Math.round``.

    It rounds half *up*, including for negatives, where Python's ``round`` uses
    banker's rounding. Every quantization here has to agree with the reference
    cell for cell, so the tie-break is spelled out rather than inherited.

    NaN and the infinities pass through, as they do in JavaScript. Python's
    ``math.floor`` raises on them, and this function is reached from the
    drawing paths, where the reference's answer is "draw nothing" rather than
    "crash the app".
    """
    if value != value or math.isinf(value):
        return value
    return math.floor(value + 0.5)


class Color(int):
    """A packed terminal color. Arithmetic on it is meaningless; use the methods."""

    __slots__ = ()

    def __repr__(self) -> str:  # pragma: no cover - debugging aid
        if self == 0:
            return "Color.DEFAULT"
        if self & _IDX_FLAG:
            return f"Color.ansi256({self & 255})"
        return f"Color.hex(0x{self & 0xFFFFFF:06x})"

    # ------------------------------------------------------------ construction

    @staticmethod
    def rgb(r: int, g: int, b: int) -> "Color":
        """Truecolor from 0-255 components. Values outside the range wrap, as
        the reference's ``& 255`` does."""
        return Color(_RGB_FLAG | (r & 255) << 16 | (g & 255) << 8 | (b & 255))

    @staticmethod
    def hex(value: "str | int") -> "Color":
        """``Color.hex("#00d7ff")``, ``Color.hex("#0df")`` or ``Color.hex(0x00d7ff)``.

        Anything unparseable becomes black rather than an error: a color is
        cosmetic, and a theme that fails to load is worse than one that is wrong.
        """
        if isinstance(value, int):
            return Color(_RGB_FLAG | (value & 0xFFFFFF))
        s = value.strip().lstrip("#")
        if len(s) == 3:
            s = s[0] * 2 + s[1] * 2 + s[2] * 2
        try:
            n = int(s, 16)
        except ValueError:
            n = 0
        return Color(_RGB_FLAG | (n & 0xFFFFFF))

    @staticmethod
    def ansi256(index: int) -> "Color":
        """An explicit xterm-256 palette entry. Rarely needed; truecolor is
        quantized for you when the terminal cannot do better."""
        return Color(_IDX_FLAG | (index & 255))

    # -------------------------------------------------------------- inspection

    @property
    def is_default(self) -> bool:
        return self == 0

    @property
    def _indexed(self) -> bool:
        return bool(self & _IDX_FLAG)

    @property
    def r(self) -> int:
        return (self >> 16) & 255

    @property
    def g(self) -> int:
        return (self >> 8) & 255

    @property
    def b(self) -> int:
        return self & 255

    # ----------------------------------------------------------------- blending

    def mix(self, other: "Color", t: float) -> "Color":
        """Blend towards ``other``. ``t`` of 0 returns self, 1 returns other.

        Software alpha — terminals have none.
        """
        if self.is_default or other.is_default:
            return self if t < 0.5 else other
        k = 0.0 if not t > 0 else (1.0 if t > 1 else t)
        return Color.rgb(
            int(round_half_up(self.r + (other.r - self.r) * k)),
            int(round_half_up(self.g + (other.g - self.g) * k)),
            int(round_half_up(self.b + (other.b - self.b) * k)),
        )

    def alpha(self, background: "Color", a: float) -> "Color":
        """Blend over a background at ``a`` (0-1), for subtle fills and shadows."""
        return background.mix(self, a)

    def lighten(self, amount: float = 0.2) -> "Color":
        return self.mix(Color.rgb(255, 255, 255), amount)

    def darken(self, amount: float = 0.2) -> "Color":
        return self.mix(Color.rgb(0, 0, 0), amount)

    def luminance(self) -> float:
        """Relative luminance, 0-1."""

        def channel(v: int) -> float:
            x = v / 255
            return x / 12.92 if x <= 0.03928 else ((x + 0.055) / 1.055) ** 2.4

        return 0.2126 * channel(self.r) + 0.7152 * channel(self.g) + 0.0722 * channel(self.b)

    def contrast(self, other: "Color") -> float:
        """WCAG contrast ratio against another color (1-21)."""
        a, b = self.luminance(), other.luminance()
        return (max(a, b) + 0.05) / (min(a, b) + 0.05)

    def grayscale(self) -> "Color":
        """Desaturate towards grey — this powers monochrome mode."""
        if self.is_default:
            return self
        v = int(round_half_up(0.299 * self.r + 0.587 * self.g + 0.114 * self.b))
        return Color.rgb(v, v, v)

    # -------------------------------------------------------------- quantizing

    def to_256(self) -> int:
        """Quantize to the xterm-256 palette, for terminals without truecolor."""
        if self._indexed:
            return self & 255
        r, g, b = self.r, self.g, self.b
        # The grey ramp often beats the cube for desaturated colors.
        if abs(r - g) < 8 and abs(g - b) < 8:
            if r < 8:
                return 16
            if r > 248:
                return 231
            return 232 + int(round_half_up((r - 8) / 247 * 24))
        return 16 + 36 * _nearest_cube(r) + 6 * _nearest_cube(g) + _nearest_cube(b)

    def to_16(self) -> int:
        """Quantize to the 16-color palette, for last-resort terminals."""
        if self._indexed:
            i = self & 255
            return i if i < 16 else from_256(i).to_16()
        r, g, b = self.r, self.g, self.b
        best, best_d = 7, None
        for i, (br, bg, bb) in enumerate(_BASE16):
            d = (r - br) ** 2 + (g - bg) ** 2 + (b - bb) ** 2
            if best_d is None or d < best_d:
                best, best_d = i, d
        return best


DEFAULT_COLOR = Color(0)
"""The terminal's own foreground/background. Never emitted as an SGR color."""

# Module-level aliases, so callers can write `rgb(1, 2, 3)` as the reference does.
rgb = Color.rgb
hex_color = Color.hex
ansi256 = Color.ansi256


def _nearest_cube(v: int) -> int:
    best, best_d = 0, None
    for i, c in enumerate(_CUBE):
        d = abs(c - v)
        if best_d is None or d < best_d:
            best, best_d = i, d
    return best


def from_256(index: int) -> Color:
    """Convert a 256-palette index back to truecolor."""
    i = index & 255
    if i < 16:
        r, g, b = _BASE16[i]
        return Color.rgb(r, g, b)
    if i >= 232:
        v = 8 + (i - 232) * 10
        return Color.rgb(v, v, v)
    n = i - 16
    return Color.rgb(_CUBE[(n // 36) % 6], _CUBE[(n // 6) % 6], _CUBE[n % 6])


class Gradient:
    """A multi-stop color ramp.

    ::

        heat = Gradient([Color.hex("#00d7ff"), Color.hex("#ff5f5f")])
        warm = heat.sample(0.75)
    """

    __slots__ = ("stops",)

    def __init__(self, stops: Iterable[Color] = ()) -> None:
        self.stops: Sequence[Color] = tuple(stops)

    def sample(self, t: float) -> Color:
        n = len(self.stops)
        if n == 0:
            return DEFAULT_COLOR
        if n == 1:
            return self.stops[0]
        # Ordered so NaN falls through to 0 rather than indexing wild.
        k = 1.0 if t > 1 else (t if t > 0 else 0.0)
        pos = k * (n - 1)
        i = min(int(math.floor(pos)), n - 2)
        return self.stops[i].mix(self.stops[i + 1], pos - i)

    def steps(self, n: int) -> list[Color]:
        """Sample ``n`` evenly spaced colors."""
        return [self.sample(0.0 if n == 1 else i / (n - 1)) for i in range(n)]
